WordPress/gutenberg
# Blocks: Adding versioning and migrations to the Block API

refs #2541 

This PR does not solve all migrations paths. It only addresses the **"automatic" migration on post edit** strategy.

This is now ready for a first review. For testing purpose, I've upgraded the quote block to use `cite` instead of `footer` with a migration.

**How does it work**

This implements the parsing flow described here https://github.com/WordPress/gutenberg/issues/2541#issuecomment-341671584

- All blocks are assumed to be in version `1` by default
- You can increase the version of the block using a `version` property and the version is persisted in the block starting comment
- If you do so, you'd more likely want to add a "migration" under the `migrations` property.
- A migration consists of the old `attributes`, the old `save` function and the version to upgrade.
- When parsing a block, if we detect that it's on an old version and an automatic migration exists, we apply it.

**Notes**

 - I noticed that saving the quote block with a `cite` and reloading the page adds some undesired `<p>` tags which causes the parsing to fail. I wonder if it's something related to `autop` that was not causing an issue when it was a `footer` tag instead.

 - I added a "version" to the block starting comment, but this revealed an issue with the `Block_Attributes` parser. 

**Todo:**

 - [ ] Fix the `autop` behavior breaking `cite` 
 - [x] Fix the attributes as JSON parsing

**Testing instructions**

 - Paste this in to the code editor (old quote HTML)

```html

<blockquote class="wp-block-quote blocks-quote-style-1">
    <p>The editor will endeavour to create a new page and post building experience that makes writing rich posts effortless, and has “blocks” to make it easy what today might take shortcodes, custom HTML, or “mystery meat” embed discovery.</p><footer>Matt Mullenweg, 2017</footer></blockquote>

```

 - Click outside the editable area
 - You'll notice the block is "migrated" to the last version with `cite`
